src/hev-config.c

```diff
@@ -28,6 +31,72 @@ static int read_write_timeout = 60000;
 static int limit_nofile = -2;
 static int log_level = HEV_LOGGER_WARN;
 
+static int
+hev_config_parse_server (yaml_document_t *doc, yaml_node_t *base, const char *sec,
+                       HevConfigServer *srv)
+{
+    yaml_node_pair_t *pair;
+    const char *addr = NULL;
+    const char *port = NULL;
+    const char *login = NULL;
+    const char *password = NULL;
+
+    if (!base || YAML_MAPPING_NODE != base->type || !srv)
+        return -1;
+
+    for (pair = base->data.mapping.pairs.start;
+         pair < base->data.mapping.pairs.top; pair++) {
+        yaml_node_t *node;
+        const char *key, *value;
+
+        if (!pair->key || !pair->value)
+            break;
+
+        node = yaml_document_get_node (doc, pair->key);
+        if (!node || YAML_SCALAR_NODE != node->type)
+            break;
+        key = (const char *)node->data.scalar.value;
+
+        node = yaml_document_get_node (doc, pair->value);
+        if (!node || YAML_SCALAR_NODE != node->type)
+            break;
+        value = (const char *)node->data.scalar.value;
+
+        if (0 == strcmp (key, "port"))
+            port = value;
+        else if (0 == strcmp (key, "address"))
+            addr = value;
+        else if (0 == strcmp (key, "login"))
+            login = value;
+        else if (0 == strcmp (key, "password"))
+            password = value;
+    }
+
+    if (!port) {
+        fprintf (stderr, "Can't found %s.port!\n", sec);
+        return -1;
+    }
+
+    if (!addr) {
+        fprintf (stderr, "Can't found %s.address!\n", sec);
+        return -1;
+    }
+
+    if ((login && !password) || (!login && password)) {
+        fprintf (stderr, "Must be set both login and password %s!\n", sec);
+        return -1;
+    }
+
+    srv->addr = strdup(addr);
```

Comment:
If it run as library (e.g. JNI), I think needs to free. or use byte array and strncpy like before?

Reply:
you application never call this method more then once at launch. And this pointer used all time wile application work.
In "library style" let to user to do free() if it needed.

Reply:
See: https://github.com/heiher/hev-socks5-tproxy/blob/master/src/hev-jni.c#L80

I think the internal configs is invisible to the user.